src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/RazorComponentRenameEndpoint.cs

```diff
@@ -0,0 +1,292 @@
+        private readonly RazorComponentSearchEngine _componentSearchEngine;
+        private readonly ILogger _logger;
+
+        private RenameCapability _capability;
+
+        public RazorComponentRenameEndpoint(
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver,
+            RazorComponentSearchEngine componentSearchEngine,
+            ProjectSnapshotManagerAccessor projectSnapshotManagerAccessor,
+            ILoggerFactory loggerFactory)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher ?? throw new ArgumentNullException(nameof(foregroundDispatcher));
+            _documentResolver = documentResolver ?? throw new ArgumentNullException(nameof(documentResolver));
+            _componentSearchEngine = componentSearchEngine ?? throw new ArgumentNullException(nameof(componentSearchEngine));
+            _projectSnapshotManager = projectSnapshotManagerAccessor?.Instance ?? throw new ArgumentNullException(nameof(projectSnapshotManagerAccessor));
+            _logger = loggerFactory.CreateLogger<RazorComponentRenameEndpoint>();
+        }
+
+        public RenameRegistrationOptions GetRegistrationOptions()
+        {
+            return new RenameRegistrationOptions
+            {
+                PrepareProvider = false,
+                DocumentSelector = RazorDefaults.Selector,
+            };
+        }
+
+        public async Task<WorkspaceEdit> Handle(RenameParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            if (!FileKinds.IsComponent(codeDocument.GetFileKind()))
+            {
+                return null;
+            }
+
+            var originTagHelperBinding = await GetOriginTagHelperBindingAsync(documentSnapshot, codeDocument, request.Position).ConfigureAwait(false);
+            if (originTagHelperBinding is null)
+            {
+                return null;
+            }
+
+            var originTagDescriptor = originTagHelperBinding.Descriptors.First();
```

Comment:
Will throw if `originTagHelperBinding.Descriptors.Count == 0`, is that a possibility?